### ISSUE TITLE
Chapter Two (Linear Model)

### ISSUE THREAD

**Issue by saberomidii** [saberomidii/ML_Torch#3]

**Comment by saberomidii** [comment]
import torch
X=torch.tensor([1,2,3,4],dtype= torch.float32)
Y=torch.tensor([2,4,6,8],dtype= torch.float32)

w=torch.tensor(0.0,dtype=torch.float32,requires_grad=True)


#model prediction 
def forward(x):
  return w*x

#loss 
def loss(y,y_predicted):
  return ((y_predicted-y)**2).mean()


print(f'Prediction before training: f(5)= {forward(5):.3f}')

#training 
learning_rate = 0.01;
n_iters =61

for epoch in range(n_iters):
  #prediction =forward pass
  y_pred = forward(X)

  #loss
  l=loss(Y, y_pred)

  #gradients =backward pass
  l.backward() #dl/dw

  #update weights 
  with torch.no_grad():
    w-=learning_rate*w.grad
  

  w.grad.zero_()
  if epoch % 5 ==0:
    print(f"epoch {epoch+1}:w={w:.2f},loss={l:.5f}") 

print(f"Prediction after training: f(5) = {forward(5):.3f}")

**Comment by saberomidii** [comment]
```
#Design model 
#construct loos and optimizer
# Training loop 
#forward pass
#backward pass 
# update weights 

import torch
import torch.nn as nn 

X=torch.tensor([[1],[2],[3],[4]],dtype= torch.float32)
Y=torch.tensor([[2],[4],[6],[8]],dtype= torch.float32)


X_test = torch.tensor([5],dtype=torch.float32)

n_samples, n_features = X.shape
print(n_samples,n_features)

input_size = n_features
output_size = n_features


model = nn.Linear(input_size, output_size)

print(f'Prediction before training: f(5)= {model(X_test).item():.3f}')

#training 
learning_rate = 0.01;
n_iters =1150

#loss 

loss= nn.MSELoss()
optimizer=torch.optim.SGD(model.parameters(),lr=learning_rate)

for epoch in range(n_iters):
  #prediction =forward pass
  y_pred = model(X)

  #loss
  l=loss(Y,y_pred)
  #gradients =backward pass
  l.backward() #dl/dw

  #update weights 
  optimizer.step()

  optimizer.zero_grad()
  if epoch % 5 ==0:
    [w,b]=model.parameters()
    print(f"epoch {epoch+1}:w={w[0][0].item():.2f},loss={l:.5f}") 

print(f'Prediction before training: f(5)= {model(X_test).item():.3f}')

```
